prime_server/prime_server.hpp

```diff
@@ -59,8 +60,27 @@ constexpr uint32_t DEFAULT_REQUEST_TIMEOUT = -1;              // infinity second
 // TODO: bundle both request_containter_t (req, rep) and request_info_t into
 // a single session_t that implements all the guts of the protocol
 
+// Struct that stores the a vector of shortcircuiter functions and its priorities
+template <class request_container_t>
+struct shortcircuiters_t {
```

Comment:
i dont think we need a `struct` to encapsulate this idea. at least not in the public (header) interface. remember that in public interfaces we want to expose as little as possible. the more we put here the more we have to maintain.

all we need here is a single type alias for the functor:

`using shortcircuiter_t = std::function<std::unique_ptr<zmq::message_t>(const request_container_t&)>;`

then its up to any protocol to provide that optionally. so over in each protocols header you'd make a free function to return the `shortcircuiter_t` object. each protocol can have different arguments to this shortcircuiter generator function. http will want healthcheck path and options mask (with defaults) whereas netstring will just want healthcheck.

then you can hide the actual implementation of this inside of http_protocol.cpp and netstring_protocol.cpp respectively.

no need for keeping structs around or a complex interface. you just call a function from each protocol that returns a lambda that does the work. you dont even need vectors of shortcircuiters. each protocol just has their priority list hardcoded i suppose (probably makes sense since protocols have a protocol hahaha).

this might be a bit of an advance c++ pattern and im not sure how familiar you are with that so please let me know if any of this is making sense. if not i can write some sample code to show what i mean.

Reply:
I think with my last few commits i'm closer to that idea!!